src/BigDecimal.ts

```diff
@@ -0,0 +1,759 @@
+    return this.toJSON()
+  },
+  pipe() {
+    return pipeArguments(this, arguments)
+  }
+} as const
+
+/**
+ * Checks if a given value is a `BigDecimal`.
+ *
+ * @param u - The value to check.
+ *
+ * @since 2.0.0
+ * @category guards
+ */
+export const isBigDecimal = (u: unknown): u is BigDecimal => typeof u === "object" && u !== null && TypeId in u
+
+/**
+ * Creates a `BigDecimal` from a `bigint` value and a scale.
+ *
+ * @param value - The `bigint` value to create a `BigDecimal` from.
+ * @param scale - The scale of the `BigDecimal`.
+ *
+ * @since 2.0.0
+ * @category constructors
+ */
+export const scaled = (value: bigint, scale: number): BigDecimal => {
+  const o = Object.create(BigDecimalProto)
+  o.value = value
+  o.scale = scale
+  return o
+}
+
+/**
+ * Creates a `BigDecimal` from a `bigint` or `number` value.
+ *
+ * @param value - The `bigint` or `number` value to create a `BigDecimal` from.
+ *
+ * @since 2.0.0
+ * @category constructors
+ */
+export const make = (value: bigint | number): BigDecimal => {
+  if (typeof value === "number") {
+    const [lead, trail = ""] = `${value}`.split(".")
+    return scaled(BigInt(`${lead}${trail}`), trail.length)
+  }
+
+  return scaled(value, 0)
+}
+
+const bigint0 = BigInt(0)
+const bigint10 = BigInt(10)
+const zero = make(bigint0)
+
+/**
+ * Parses a numerical `string` into a `BigDecimal`.
+ *
+ * @param s - The `string` to parse.
+ *
+ * @example
+ * import { parse, make } from 'effect/BigDecimal'
+ * import { some, none } from 'effect/Option'
+ *
+ * assert.deepStrictEqual(parse('123'), some(make(123n)))
+ * assert.deepStrictEqual(parse('123.456'), some(make(123.456)))
+ * assert.deepStrictEqual(parse('123.abc'), none())
+ *
+ * @since 2.0.0
+ * @category constructors
+ */
+export const parse = (s: string): Option.Option<BigDecimal> => {
+  let digits: string
+  let scale: number
+
+  const dot = s.search(/\./)
+  if (dot !== -1) {
+    const lead = s.slice(0, dot)
+    const trail = s.slice(dot + 1)
+    digits = `${lead}${trail}`
+    scale = trail.length
+  } else {
+    digits = s
+    scale = 0
+  }
+
+  if (digits === "") {
+    // TODO: This mimics the BigInt constructor behavior. Should this be `Option.none()`?
+    return Option.some(zero)
+  }
+
+  if (!/^(?:\+|-)?\d+$/.test(digits)) {
+    return Option.none()
+  }
+
+  return Option.some(scaled(BigInt(digits), scale))
+}
+
+/**
+ * Normalizes a given `BigDecimal` by removing trailing zeros.
+ *
+ * @param self - The `BigDecimal` to normalize.
+ *
+ * @example
+ * import { normalize, make } from 'effect/BigDecimal'
+ *
+ * assert.deepStrictEqual(normalize(make(123.456)), make(123.456))
+ * assert.deepStrictEqual(normalize(make(123.456000)), make(123.456))
+ * assert.deepStrictEqual(normalize(make(123.000456)), make(123.000456))
+ * assert.deepStrictEqual(normalize(make(123.000000)), make(123))
+ *
+ * @since 2.0.0
+ * @category constructors
+ */
+export const normalize = (self: BigDecimal): BigDecimal => {
```

Comment:
We could probably cache the normalized result inside the BigDecimal structure.

That should improve performance of the Hash implementation.

Reply:
You mean in the original `self` in case `normalize` is called on it again? Yeah that's clever!